modules/cudev/test/test_nd.cu

```diff
@@ -0,0 +1,198 @@
+#include "test_precomp.hpp"
+
+using namespace cv;
+using namespace cv::cuda;
+using namespace cvtest;
+
+using ElemTypes = ::testing::Types<
+    Vec<uchar, 1>, Vec<uchar, 2>, Vec<uchar, 3>, Vec<uchar, 4>, // CV_8U
+    Vec<schar, 1>, Vec<schar, 2>, Vec<schar, 3>, Vec<schar, 4>, // CV_8S
+    Vec<ushort, 1>, Vec<ushort, 2>, Vec<ushort, 3>, Vec<ushort, 4>, // CV_16U
+    Vec<short, 1>, Vec<short, 2>, Vec<short, 3>, Vec<short, 4>, // CV_16S
+    Vec<int, 1>, Vec<int, 2>, Vec<int, 3>, Vec<int, 4>, // CV_32S
+    Vec<float, 1>, Vec<float, 2>, Vec<float, 3>, Vec<float, 4>, // CV_32F
+    Vec<double, 1>, Vec<double, 2>, Vec<double, 3>, Vec<double, 4>, //CV_64F
+    Vec<float16_t, 1>, Vec<float16_t, 2>, Vec<float16_t, 3>, Vec<float16_t, 4> // CV_16F
+>;
+
+template <typename ElemType>
+class GpuMatNDTest : public ::testing::Test
+{
+public:
+    using MatType = Mat_<ElemType>;
+    using CnType = typename Mat_<ElemType>::channel_type;
+    static constexpr int cn = DataType<ElemType>::channels;
+    using SizeArray = GpuMatND::SizeArray;
+
+    static void doTest(const SizeArray& size)
+    {
+        const int dims = static_cast<int>(size.size());
+
+        MatType gold(dims, size.data());
+        MatType dst;
+
+        // init gold with random data
+        for (ElemType& elem : gold)
+            for (int i = 0; i < cn; ++i)
+                elem[i] = static_cast<CnType>(std::rand());
+
+        GpuMatND gmat;
+
+        {
+            // simple upload, download test for GpuMatND
+
+            gmat.upload(gold);
+            gmat.download(dst);
+
+            EXPECT_TRUE(std::equal(gold.begin(), gold.end(), dst.begin()));
+        }
```

Comment:
It'll be great if you split the blocks like this on dedicated parameterized tests with GTest. It simplifies issue triaging and reduces side effects if one block affects another during execution. The test you used as base was implemented in prehistoric period and we try to get rid of large 'doTest' things that does everything.

Reply:
Currently, it uses type-parameterized tests. Should I change it to a value parameterized form?

Reply:
Up to you.

Reply:
I have fixed it.